### PR TITLE
Adding a new task to download GitHub packages

### PR DESCRIPTION
**Pull request by romil07**

### CODE REVIEW

File: Tasks/DotnetCLIRestore/dotnetrestore.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import tl = require("azure-pipelines-task-lib/task");
+import tr = require("azure-pipelines-task-lib/toolrunner");
+import path = require("path");
+
+import * as restoreCommand from './restorecommand';
+
+let MessagePrinted = false;
+
+export class dotNetExe {
```
Comment (kasubram) [inline Tasks/DotnetCLIRestore/dotnetrestore.ts]: I dont know the convention. class name with camelcasing?

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    updateXmlFile(projectFilePath, (xml: any): any => {
+        if (xml) {
+            if (xml.getName().toLowerCase() !== "project") {
+                throw Error("Expected project element");
+            }
+            let xmlPropertyGroup = getOrCreateLastElement(xml, "PropertyGroup");
+            let xmlOutputType = xmlPropertyGroup.c("OutputType");
+            let xmlTargetFramework = xmlPropertyGroup.c("TargetFramework");
+            let xmlOutputTypeTxt = xmlOutputType.t("Exe");
+            let xmlTargetFrameworkTxt = xmlTargetFramework.t("netcoreapp2.1");
```
Comment (hiyadav) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: Given that now 3.1 is the latest LTS version, should we rather use that ?

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    const dotnet = tl.tool(dotnetPath);
+    dotnet.arg('add');
+
+    if (projectFile) {
+        dotnet.arg(projectFile);
+    }
+
+    if (packagesDirectory) {
+        dotnet.arg('package');
+        dotnet.arg('OctocatApp');
```
Comment (hiyadav) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: will it always be octocatapp ?

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+function dotNetRestoreAsync(dotnetPath: string, projectFile: string, packagesDirectory: string, version: string, configFile: string, noCache: boolean, verbosity: string): Q.Promise<number> {
+    const dotnet = tl.tool(dotnetPath);
+    dotnet.arg('add');
+
+    if (projectFile) {
+        dotnet.arg(projectFile);
+    }
+
+    if (packagesDirectory) {
```
Comment (hiyadav) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: shouldn't the below code be independent of packagesDirectory being present ?

File: Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts [hunk 5]
```diff
@@ -144,14 +144,16 @@ export class NuGetConfigHelper2 {
 
     private getTempNuGetConfigPath(): string {
         const tempNuGetConfigBaseDir = NuGetConfigHelper2.getTempNuGetConfigBasePath();
-        const tempNuGetConfigFileName = "tempNuGet_" + tl.getVariable("build.buildId") + ".config";
+        // Changing this to standard name of nuget.config. In one of the tasks, dotnet add command is being used and
+        // add command does not take the name of the config file as parameter. It assumes default name.
+        // This won't impact anything as this file is anyway deleted right after the operation.
+        const tempNuGetConfigFileName =  "nuget.config";
```
Comment (hiyadav) [inline Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts]: This will not bode well with cases where, user specifies a nugetConfigPath which has nuget.config already present. Like in DotNetCore CLI task and might be other as well

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            nuGetConfigPath = nuGetConfigHelper.tempNugetConfigPath;
+        } else {
+            tl.debug('No sources were added to the temp NuGet.config file');
+        }
+
+        // Setting creds in the temp NuGet.config if needed
+        nuGetConfigHelper.setAuthForSourcesInTempNuGetConfig();
+        const configFile = nuGetConfigHelper.tempNugetConfigPath;
+        nuGetConfigHelper.backupExistingRootNuGetFiles();
+        const dotnetPath = tl.which('dotnet', true);
```
[the remark this answers is not in the document]
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Removed this.

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        if (!tl.filePathSupplied('packagesDirectory')) {
+            packagesDirectory = null;
+        }
+
+        const externalAuthArr: auth.ExternalAuthInfo[] = GetExternalAuthInfoArray('externalEndpoints', username);
+        const authInfo = new auth.NuGetExtendedAuthInfo(null, externalAuthArr);
+
+        // Setting up sources, either from provided config file or from feed selection
+        tl.debug('Setting up sources');
+        let nuGetConfigPath: string = undefined;
```
[the remark this answers is not in the document]
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Removed this.

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+        tl.setResult(tl.TaskResult.Succeeded, tl.loc('PackagesInstalledSuccessfully'));
+
+    } catch (err) {
+        tl.error(err);
+        if (buildIdentityDisplayName || buildIdentityAccount) {
+            tl.warning(tl.loc('BuildIdentityPermissionsHint', buildIdentityDisplayName, buildIdentityAccount));
+        }
+
+        tl.setResult(tl.TaskResult.Failed, tl.loc('PackagesFailedToInstall'));
```
[the remark this answers is not in the document]
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: No. I am assuming you are concerned with the "plural" packages in the message. Changing this.

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        xmlString = xmlString.substr(1);
+    }
+
+    let xml = ltx.parse(xmlString);
+    xml = updateFn(xml);
+    fs.writeFileSync(xmlPath, xml.root().toString());
+}
+
+function getProjectFilePath(): string {
+    const tempNuGetConfigBaseDir = tl.getVariable("Agent.BuildDirectory") || tl.getVariable("Agent.TempDirectory");
```
[the remark this answers is not in the document]
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Since the base path is Agent.BuildDirectory and I don't expect that to change, so it shouldn't matter.
But I included this anyway.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 3 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: This is a test code that crept in. Removed it.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 2 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: I tried using 3.0 but was getting some error. So, I will stick to this.

File: Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts (hunk 5 above)
Comment (romil07) [reply Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts]: I will discuss this with you.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 4 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: This is a user input. Packages will be downloaded in this directory.

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+        if (buildIdentityDisplayName || buildIdentityAccount) {
+            tl.warning(tl.loc('BuildIdentityPermissionsHint', buildIdentityDisplayName, buildIdentityAccount));
+        }
+
+        tl.setResult(tl.TaskResult.Failed, tl.loc('PackageFailedToInstall'));
+    }
+}
+
+function getTempNuGetConfigPath(): string {
+    const tempNuGetConfigBaseDir = tl.getVariable("Agent.BuildDirectory") || tl.getVariable("Agent.TempDirectory");
```
Comment (hiyadav) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: Nitpick: Should we rather prioritize use us tempDirectory since the file name here is NuGet.config

File: Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts (hunk 5 above)
Comment (romil07) [reply Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts]: Closing this with the updated design.

File: Tasks/Common/utility-common/package-lock.json [hunk 11]
```diff
@@ -9,7 +9,7 @@
       "resolved": "https://registry.npmjs.org/argparse/-/argparse-1.0.10.tgz",
```
Comment (ds-ms) [inline Tasks/Common/utility-common/package-lock.json]: Undo these changes

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        dotnet.arg(version);
+    }
+
+    dotnet.arg('-n');
+
+    const envWithProxy = ngRunner.setNuGetProxyEnvironment(process.env, configFile, null);
+    return dotnet.exec({ cwd: path.dirname(projectFile), env: envWithProxy } as IExecOptions);
+}
+
+function dotNetRestoreAsync(dotnetPath: string, projectFile: string, packagesDirectory: string, version: string, configFile: string, noCache: boolean, verbosity: string): Q.Promise<number> {
```
[the remark this answers is not in the document]
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Right.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 10 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Keeping it as it is.

File: Tasks/Common/utility-common/package-lock.json (hunk 11 above)
Comment (romil07) [reply Tasks/Common/utility-common/package-lock.json]: Done.

File: Tasks/DotnetCLIRestore/task.json [hunk 13]
```diff
@@ -0,0 +1,159 @@
+{
+    "id": "9fac244b-8d7c-4d8e-a003-2097daa3270f",
+    "name": "DownloadGitHubNugetPackage",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: Better use same name for folder as well

File: Tasks/DotnetCLIRestore/task.json [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    "friendlyName": "Download Nuget GitHub Packages",
+    "description": "Restore your nuget packages using dotnet CLI",
+    "author": "Microsoft Corporation",
+    "helpUrl": "https://docs.microsoft.com/azure/devops/pipelines/tasks/build/dotnet-core-cli",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: I hope this will be changed

File: Tasks/DotnetCLIRestore/task.json [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    "helpMarkDown": "[Learn more about this task](https://go.microsoft.com/fwlink/?linkid=832194) or [see the .NET Core documentation](https://docs.microsoft.com/dotnet/core/)",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: this as well

File: Tasks/DotnetCLIRestore/task.json [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+        "Agent"
+    ],
+    "demands": [],
+    "version": {
+        "Major": 1,
+        "Minor": 0,
+        "Patch": 0
+    },
+    "minimumAgentVersion": "2.115.0",
+    "instanceNameFormat": "dotnet $(command)",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: there is no input `command`. Fix this

File: Tasks/DotnetCLIRestore/task.json [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+            "required": true,
+            "helpMarkDown": "Version of the package to download from github"
+        },
+        {
+            "name": "externalEndpoints",
+            "aliases": [
+                "externalFeedCredentials"
+            ],
+            "type": "connectedService:github",
+            "label": "Credentials for feeds outside this organization/collection",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: wrong label which talks about ADO org/collection. this is purely github endpoint

File: Tasks/DotnetCLIRestore/task.json [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+            "required": false,
+            "helpMarkDown": "Credentials to use for external registries located in the selected NuGet.config. For feeds in this organization/collection, leave this blank; the build’s credentials are used automatically.",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: this as well

File: Tasks/DotnetCLIRestore/task.json [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+            "helpMarkDown": "Version of the package to download from github"
+        },
+        {
+            "name": "externalEndpoints",
+            "aliases": [
+                "externalFeedCredentials"
+            ],
+            "type": "connectedService:github",
+            "label": "Credentials for feeds outside this organization/collection",
+            "required": false,
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: what credential is used if endpoint is not specified

File: Tasks/DotnetCLIRestore/task.json [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+        },
+        {
+            "name": "packagesDirectory",
+            "aliases": [
+                "restoreDirectory"
+            ],
+            "type": "string",
+            "label": "Destination directory",
+            "defaultValue": "",
+            "helpMarkDown": "Specifies the folder in which packages are installed. If no folder is specified, packages are restored into the default NuGet package cache.",
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: fallback target should be nuget cache or working directory? If I understand correctly this task will mostly be used for downloading tools etc do current working directory would be better

File: Tasks/DotnetCLIRestore/task.json [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+            "required": "false",
+            "groupName": "restoreAdvanced"
+        }
+    ],
+    "dataSourceBindings": [
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: are the two data binding required?

File: Tasks/DotnetCLIRestore/dotnetrestore.ts [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    public async execute() {
+        tl.setResourcePath(path.join(__dirname, "task.json"));
+        this.setConsoleCodePage();
+
+        try {
+            await restoreCommand.run();
+        }
+        finally {
+            if (!MessagePrinted) {
+               console.log(tl.loc('NetCore3Update'));
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/dotnetrestore.ts]: I don't think it makes sense to print this message as we are internally deciding .net core version.

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 23]
```diff
@@ -0,0 +1,306 @@
...
+    const buildIdentityDisplayName: string = null;
+    const buildIdentityAccount: string = null;
+
+    try {
+        let packageName: string = tl.getInput("packageName");
+        let packageVersion: string = tl.getInput("version") || "";
+
+        if (!packageName)
+        {
+            return;
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: it should fail with proper error if a valid package name is not specified

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+        }
+        else {
+            packageName = packageName.split("/")[1];
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: if you are assuming that package name would have a format of `<foo>/<bar>` then the name should be validated for this format

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 25]
```diff
@@ -0,0 +1,306 @@
...
+        }
+
+        let endpointNames = tl.getDelimitedInput("externalEndpoints", ',');
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: this input was optional so either make is mandatory or check for null here

File: Tasks/DotnetCLIRestore/restorecommand.ts [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+        const username: string = await GetGitHubUser(endpointNames[0]); // we will always have a single connection
+
+        const noCache = true;
+        const verbosity = "minimal";
+        let packagesDirectory = tl.getPathInput('packagesDirectory');
+        if (!tl.filePathSupplied('packagesDirectory')) {
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/restorecommand.ts]: Curious: what does this function do?

File: Tasks/DotnetCLIRestore/task.json [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+            "aliases": [
+                "externalFeedCredentials"
+            ],
+            "type": "connectedService:github",
+            "label": "Credentials for feeds outside this organization/collection",
+            "required": false,
+            "helpMarkDown": "Credentials to use for external registries located in the selected NuGet.config. For feeds in this organization/collection, leave this blank; the build’s credentials are used automatically.",
+            "properties": {
+                "EditableOptions": "False",
+                "MultiSelectFlatList": "True"
```
Comment (bishal-pdMSFT) [inline Tasks/DotnetCLIRestore/task.json]: Curious to know the use case of selecting multiple github connections

File: Tasks/DotnetCLIRestore/task.json (hunk 14 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/task.json]: Since I don't have a ready doc yet, I will change this to the generic doc page for tasks.

File: Tasks/DotnetCLIRestore/task.json (hunk 15 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/task.json]: Same as above.

File: Tasks/DotnetCLIRestore/task.json (hunk 19 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/task.json]: It has to be specified. Marked this true.

File: Tasks/DotnetCLIRestore/task.json (hunk 20 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/task.json]: That's what our logic is. Somehow I missed the documentation of individual inputs. I am updating this here.
Thanks for looking at it closely.

File: Tasks/DotnetCLIRestore/task.json (hunk 21 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/task.json]: I don't have a use of them but looks like they are mandatory. I couldn't compile without these 2.

File: Tasks/DotnetCLIRestore/dotnetrestore.ts (hunk 22 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/dotnetrestore.ts]: I copied this from CLIV2 task. If you think it shouldn't be displayed, I will remove it. Let me know.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 25 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Made this a mandatory input.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 26 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: From the doc: "Checks whether a path inputs value was supplied by the user"

File: Tasks/DotnetCLIRestore/task.json (hunk 27 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/task.json]: There isn't one. Will change this.

File: Tasks/DotnetCLIRestore/restorecommand.ts (hunk 24 above)
Comment (romil07) [reply Tasks/DotnetCLIRestore/restorecommand.ts]: Added the check in if condition.

File: Tasks/DotnetCLIRestore/package.json [hunk 28]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "dotnetclirestore",
```
[the remark this answers is not in the document]
Comment (romil07) [reply Tasks/DotnetCLIRestore/package.json]: I didn't get this. Can you please explain more?

File: Tasks/DownloadGitHubNugetPackage/restorecommand.ts [hunk 29]
```diff
@@ -0,0 +1,301 @@
...
+            null,
+            authInfo,
+            { credProviderFolder: null, extensionsDisabled: true },
+            getTempNuGetConfigPath() /* tempConfigPath */,
+            false /* useNugetToModifyConfigFile */);
+
+        let credCleanup = () => { return; };
+
+        const sources: Array<auth.IPackageSource> = new Array<auth.IPackageSource>();
+        let feedUri: string = "https://nuget.pkg.github.com/" + username + "/index.json";
```
Comment (admin-simeon) [inline Tasks/DownloadGitHubNugetPackage/restorecommand.ts:56]: This is just messed up logic
